lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -1275,17 +1279,15 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
   // Check whether we already know that these two declarations are not
   // structurally equivalent.
   if (Context.NonEquivalentDecls.count(
-          std::make_pair(D1->getCanonicalDecl(), D2->getCanonicalDecl())))
+          std::make_pair(D1, D2)))
     return false;
 
-  // Determine whether we've already produced a tentative equivalence for D1.
-  Decl *&EquivToD1 = Context.TentativeEquivalences[D1->getCanonicalDecl()];
-  if (EquivToD1)
-    return EquivToD1 == D2->getCanonicalDecl();
+  // If we already had encountered these declarations, finish the recursion.
+  if (Context.Visited.count({D1, D2}))
```

Comment:
I think I should rather user std::make_pair, otherwise gcc 4.8 will break my build.